### PR TITLE
[ENHANCEMENT] Use pydantic for chat field in server and SDK

### PR DESCRIPTION
**Pull request by burtenshaw**
# Description

This PR use pydantic in the server and sdk to validate schema of chat field values.

**Type of change**

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- Refactor (change restructuring the codebase without changing functionality)
- Improvement (change adding some improvement to an existing functionality)
- Documentation update

**How Has This Been Tested**

**Checklist**

- I added relevant documentation
- I followed the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature works
- I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)

### CODE REVIEW

File: argilla-server/src/argilla_server/api/schemas/v1/chat.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+class ChatMessage(BaseModel):
+    role: StrictStr
+    content: StrictStr
+
+    @validator("role")
+    def validate_role(cls, value):
+        # no spaces allowed
+        if any([c.isspace() for c in value]):
+            raise ValueError("Role must not contain spaces")
```
Comment (jfcalvo) [inline argilla-server/src/argilla_server/api/schemas/v1/chat.py]: You can use a pydantic `Field` with a regular expression to validate this instead of a custom validation.

File: argilla-server/src/argilla_server/api/schemas/v1/chat.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        if len(value) > MAX_ROLE_LENGTH:
+            raise ValueError(f"Role must be less than {MAX_ROLE_LENGTH} characters")
```
Comment (jfcalvo) [inline argilla-server/src/argilla_server/api/schemas/v1/chat.py]: You can use a `Field` with a `max_length` attribute instead of a custom validation.

File: argilla-server/src/argilla_server/api/schemas/v1/chat.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        return value
+
+    @validator("content")
+    def validate_content(cls, value):
+        if len(value) > MAX_MESSAGE_LENGTH:
+            raise ValueError(f"Content must be less than {MAX_MESSAGE_LENGTH} characters")
+        return value
```
Comment (jfcalvo) [inline argilla-server/src/argilla_server/api/schemas/v1/chat.py]: Something similar with this one, you can use a `Field` with `max_length`.

File: argilla-server/src/argilla_server/api/schemas/v1/records.py [hunk 4]
```diff
@@ -82,13 +83,20 @@ class Config:
 
 
 class RecordCreate(BaseModel):
-    fields: Dict[str, Union[StrictStr, None, List[Dict[StrictStr, StrictStr]]]]
+    fields: Dict[str, Union[StrictStr, None, List[ChatMessage]]]
```
Comment (jfcalvo) [inline argilla-server/src/argilla_server/api/schemas/v1/records.py]: Be careful with the order here. Pydantic will parse this in order until one matches. So it's recomendable to go from more specific to less specific, so something like this:

```suggestion
    fields: Dict[str, Union[List[ChatMessage], StrictStr, None]]
```

File: argilla-server/src/argilla_server/api/schemas/v1/chat.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+from typing import List
+
+from argilla_server.pydantic_v1 import BaseModel, StrictStr, validator
+
+MAX_MESSAGE_LENGTH = 5000
+MAX_MESSAGE_COUNT = 1000
+MAX_ROLE_LENGTH = 20
+
+
+class ChatMessage(BaseModel):
```
Comment (jfcalvo) [inline argilla-server/src/argilla_server/api/schemas/v1/chat.py]: Not sure if something like `ChatFieldValue` could describe better this class:

```suggestion
class ChatFieldValue(BaseModel):
```

File: argilla-server/src/argilla_server/api/schemas/v1/records.py [hunk 6]
```diff
@@ -82,13 +83,20 @@ class Config:
...
     responses: Optional[List[UserResponseCreate]]
     suggestions: Optional[List[SuggestionCreate]]
     vectors: Optional[Dict[str, List[float]]]
 
+    @validator("fields")
+    def validate_chat_messages(cls, fields):
+        for key, value in fields.items():
+            if isinstance(value, list) and all(isinstance(item, ChatMessage) for item in value):
+                fields[key] = [msg.dict() for msg in value]
+        return fields
```
Comment (jfcalvo) [inline argilla-server/src/argilla_server/api/schemas/v1/records.py]: Why is this necessary?

File: argilla-server/src/argilla_server/api/schemas/v1/records.py [hunk 7]
```diff
@@ -148,7 +156,7 @@ class RecordUpdateWithId(RecordUpdate):
 
 class RecordUpsert(RecordCreate):
     id: Optional[UUID]
-    fields: Optional[Dict[str, Union[StrictStr, None, List[Dict[StrictStr, StrictStr]]]]]
+    fields: Optional[Dict[str, Union[StrictStr, None, List[ChatMessage]]]]
```
Comment (jfcalvo) [inline argilla-server/src/argilla_server/api/schemas/v1/records.py]: The same here, be careful with the order.

File: argilla/src/argilla/_models/_record/_record.py [hunk 8]
```diff
@@ -70,3 +82,25 @@ def validate_metadata(cls, metadata: Union[List[MetadataModel], dict]) -> List[M
...
+                        raise ValueError("Chat field values must contain 'role' and 'content' keys.")
+                    if not all(key in ["role", "content"] for key in message.keys()):
+                        warnings.warn(
+                            "Chat field values should only contain 'role' and 'content' keys. Other keys will be ignored by Argilla."
+                        )
+                        message = {key: value for key, value in message.items() if key in ["role", "content"]}
+                        validated_chat_field_values.append(ChatFieldValue(**message))
+                field_value = validated_chat_field_values
+            validated_fields[field_name] = field_value
+        return validated_fields
```
Comment (jfcalvo) [inline argilla/src/argilla/_models/_record/_record.py]: This part is a little bit difficult to read. I would try to do some refactor, maybe moving things to other validations or functions?

**Review by jfcalvo — CHANGES_REQUESTED** [review]

File: argilla-server/src/argilla_server/api/schemas/v1/chat.py [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+
+from argilla_server.pydantic_v1 import BaseModel, Field
+
+MAX_MESSAGE_LENGTH = 5000
+MAX_MESSAGE_COUNT = 1000
+MAX_ROLE_LENGTH = 20
+
+
+class ChatFieldValue(BaseModel):
+    role: str = Field(..., max_length=MAX_ROLE_LENGTH, regex=r"^\S+$")
```
Comment (jfcalvo) [inline argilla-server/src/argilla_server/api/schemas/v1/chat.py]: I would move the regex to be a constant like `MAX_ROLE_REGEX`.

Comment (jfcalvo) [reply argilla-server/src/argilla_server/api/schemas/v1/chat.py]: What about adding a `min_length` also? Even if it's covered by the regex so we are more explicit about the behavior.

File: argilla-server/src/argilla_server/api/schemas/v1/chat.py [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+    content: str = Field(..., max_length=MAX_MESSAGE_LENGTH)
```
Comment (jfcalvo) [inline argilla-server/src/argilla_server/api/schemas/v1/chat.py]: I would add a `min_length` (of `1` character?) here as well.

File: argilla-server/src/argilla_server/api/schemas/v1/records.py [hunk 11]
```diff
@@ -82,13 +83,21 @@ class Config:
...
     suggestions: Optional[List[SuggestionCreate]]
     vectors: Optional[Dict[str, List[float]]]
 
+    @validator("fields")
+    @classmethod
+    def validate_chat_fields(cls, fields):
+        for key, value in fields.items():
+            if isinstance(value, list) and all(isinstance(item, ChatFieldValue) for item in value):
+                fields[key] = [msg.dict() for msg in value]
+        return fields
```
Comment (jfcalvo) [inline argilla-server/src/argilla_server/api/schemas/v1/records.py]: I guess you are doing this so it's serializable when we are storing the record.

Instead of doing this we usually use `jsonable_encoder` on the context functions so we can still have `ChatFieldValue` instances (instead of dicts) when the data is parsed in the handler.

An example of doing this is inside the `create_response` context function:

```python
        response = await Response.create(
            db,
            values=jsonable_encoder(response_create.values),
            status=response_create.status,
            record_id=record.id,
            user_id=user.id,
            autocommit=False,
        )
```

**Review by jfcalvo — CHANGES_REQUESTED** [review]

File: argilla-server/src/argilla_server/api/schemas/v1/records.py (hunk 11 above)
Comment (burtenshaw) [reply argilla-server/src/argilla_server/api/schemas/v1/records.py]: I can't get `jsonable_encoder` to encode a `List[ChatFieldValue]` do you have a suggestion?

Comment (jfcalvo) [reply argilla-server/src/argilla_server/api/schemas/v1/records.py]: It should work if you try to execute `jsonable_encoder` to `fields` attribute. If you are not making in works I can try myself.

Comment (burtenshaw) [reply argilla-server/src/argilla_server/api/schemas/v1/records.py]: Yeah I tried that, and ran out of time.

Comment (jfcalvo) [reply argilla-server/src/argilla_server/api/schemas/v1/records.py]: No problem. I will take a look to it.

**Review by jfcalvo — APPROVED** [review]
I added some changes using `jsonable_encoder` and avoiding transforming the chat field values to dict inside the validation: https://github.com/argilla-io/argilla/pull/5450/commits/277f3140341453f55f6bcb07a1f29f71bc28969b

**Comment by burtenshaw** [comment]
> I added some changes using `jsonable_encoder` and avoiding transforming the chat field values to dict inside the validation: [277f314](https://github.com/argilla-io/argilla/commit/277f3140341453f55f6bcb07a1f29f71bc28969b)

Nice. I missed the upsert. :face_palm:
